src/audio_unit/looper.rs

```diff
@@ -0,0 +1,137 @@
+    Playing { position: usize, total: usize },
+}
+
+pub struct Looper {
+    messages: Receiver<Message>,
+    buffer: Vec<f32>,
+    state: State,
+}
+
+impl Looper {
+    pub fn new(stream_config: &StreamConfig, max_buffer_ms: u32) -> (Self, Sender<Message>) {
+        let (sender, receiver) = mpsc::channel();
+
+        let buffer = vec![0.0; util::ms_in_samples(stream_config, max_buffer_ms)];
+
+        (
+            Self {
+                messages: receiver,
+                buffer,
+                state: Off,
+            },
+            sender,
+        )
+    }
+
+    fn process_messages(&mut self) {
+        let messages: Vec<_> = self.messages.try_iter().collect();
+        for message in messages {
+            self.process_message(message);
+        }
+    }
+
+    fn process_message(&mut self, _: Message) {
+        self.state = match self.state {
+            Off => Recording { position: 0 },
+            Recording { position } => Playing {
+                position: 0,
+                total: position,
+            },
+            Playing {
+                position: _,
+                total: _,
+            } => {
+                util::zero_slice(&mut self.buffer);
+                Off
+            }
+        };
+
+        println!("looper: {:?}", self.state);
+    }
+
+    fn process_samples(&mut self, input: &[f32], output: &mut [f32]) {
+        match self.state {
+            Off => {
+                util::zero_slice(output);
+            }
+            Recording { position } => {
+                let next_position = position + input.len();
+
+                if next_position <= self.buffer.len() {
+                    self.buffer[position..next_position].copy_from_slice(input);
+
+                    if next_position == self.buffer.len() {
+                        println!("looper: out of space in the buffer. switching to playback");
+
+                        self.state = Playing {
+                            position: 0,
+                            total: next_position,
+                        };
+                    } else {
+                        self.state = Recording {
+                            position: next_position,
+                        };
+                    }
+                } else {
+                    self.process_samples_wrap_around(position, self.buffer.len(), input, output);
+                }
+            }
+            Playing { position, total } => {
+                let next_position = position + output.len();
+
+                if next_position <= total {
+                    output.copy_from_slice(&self.buffer[position..next_position]);
+
+                    self.state = Playing {
+                        position: next_position % total,
+                        total,
+                    }
+                } else {
+                    self.process_samples_wrap_around(position, total, input, output);
+                }
+            }
+        };
+    }
```

Comment:
This function is a little hard to follow (and was hard to write). Any ideas on simplifying this?

Reply:
I think it reads pretty well